File: scripts/convert_factory_presets.py
#!/usr/bin/env python3
"""Convert factory_presets.json scripts from brace to line DSL syntax."""
import json
import re
from pathlib import Path

SHAPE_MAP = {
    "sin": "sine",
    "tri": "triangle",
    "saw": "saw",
    "square": "square",
    "noise": "noise",
}


def parse_kv_body(body: str) -> dict[str, str]:
    flat = [q if q else t for q, t in re.findall(r'"([^"]+)"|(\S+)', body)]
    kv: dict[str, str] = {}
    key = None
    for item in flat:
        if key is None:
            key = item.lower()
        else:
            kv[key] = item
            key = None
    return kv


def extract_blocks(script: str) -> list[tuple[str, str]]:
    script = script.replace("\r\n", "\n")
    blocks: list[tuple[str, str]] = []
    i = 0
    while i < len(script):
        if script[i] in " \t\n":
            i += 1
            continue
        m = re.match(r"(\w+)\s*\{", script[i:])
        if not m:
            i += 1
            continue
        btype = m.group(1).lower()
        start = i + m.end() - 1
        depth = 0
        j = start
        while j < len(script):
            if script[j] == "{":
                depth += 1
            elif script[j] == "}":
                depth -= 1
                if depth == 0:
                    blocks.append((btype, script[start + 1 : j].strip()))
                    i = j + 1
                    break
            j += 1
        else:
            break
    return blocks


def convert_script(script: str) -> str:
    lines_out: list[str] = []
    counters = {"stage": 0, "filter": 0, "comp": 0, "env": 0, "osc": 0}

    for btype, body in extract_blocks(script):
        if btype == "param":
            kv = parse_kv_body(body)
            alias = kv.get("alias", "a")
            name = kv.get("name", alias)
            mn = kv.get("min", "0")
            mx = kv.get("max", "1")
            lines_out.append(f"param {alias} = {name} [{mn}, {mx}]")

        elif btype == "stage":
            counters["stage"] += 1
            name = f"stage{counters['stage']}"
            compact = " ".join(body.split())
            if compact.startswith("y "):
                formula = compact[2:].strip()
            else:
                formula = compact
            lines_out.append(f"{name}: y = {formula}")

        elif btype == "filter":
            counters["filter"] += 1
            name = f"filter{counters['filter']}"
            kv = parse_kv_body(body)
            parts = [f"{k} = {v}" for k, v in kv.items()]
            lines_out.append(f"{name}: " + "; ".join(parts))

        elif btype == "osc":
            kv = parse_kv_body(body)
            osc_name = kv.get("name", f"osc{counters['osc'] + 1}")
            counters["osc"] += 1
            shape = SHAPE_MAP.get(kv.get("shape", "sine").lower(), kv.get("shape", "sine"))
            parts = [f"shape = {shape}"]
            if "freq" in kv:
                parts.append(f"freq = {kv['freq']}")
            if "sync" in kv:
                parts.append(f"sync = {kv['sync']}")
            if "depth" in kv:
                parts.append(f"depth = {kv['depth']}")
            lines_out.append(f"{osc_name}: " + "; ".join(parts))

        elif btype == "env":
            kv = parse_kv_body(body)
            env_name = kv.get("name", f"env{counters['env'] + 1}")
            counters["env"] += 1
            mode = kv.get("mode", "rms").lower()
            parts = [f"type = {mode}"]
            if "attack" in kv:
                parts.append(f"attack = {kv['attack']}")
            if "release" in kv:
                parts.append(f"release = {kv['release']}")
            if "trigger" in kv:
                parts.append(f"trigger = {kv['trigger']}")
            lines_out.append(f"{env_name}: " + "; ".join(parts))

        elif btype == "comp":
            counters["comp"] += 1
            name = f"comp{counters['comp']}"
            kv = parse_kv_body(body)
            parts = []
            for k in ("threshold", "ratio", "attack", "release"):
                if k in kv:
                    parts.append(f"{k} = {kv[k]}")
            lines_out.append(f"{name}: " + "; ".join(parts))

    return "\n".join(lines_out)


def main() -> None:
    path = Path(__file__).resolve().parents[1] / "resources" / "factory_presets.json"
    data = json.loads(path.read_text(encoding="utf-8"))
    for preset in data:
        preset["script"] = convert_script(preset["script"])
    path.write_text(json.dumps(data, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(f"Converted {len(data)} presets -> {path}")


if __name__ == "__main__":
    main()

File: scripts/convert_factory_presets.mjs
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const shapeMap = { sin: "sine", tri: "triangle", saw: "saw", square: "square", noise: "noise" };

function parseKvBody(body) {
  const flat = [...body.matchAll(/"([^"]+)"|(\S+)/g)].map((m) => m[1] ?? m[2]);
  const kv = {};
  let key = null;
  for (const item of flat) {
    if (key === null) key = item.toLowerCase();
    else {
      kv[key] = item;
      key = null;
    }
  }
  return kv;
}

function extractBlocks(script) {
  const blocks = [];
  let i = 0;
  const s = script.replace(/\r\n/g, "\n");
  while (i < s.length) {
    if (/[\s\n]/.test(s[i])) {
      i++;
      continue;
    }
    const m = s.slice(i).match(/^(\w+)\s*\{/);
    if (!m) {
      i++;
      continue;
    }
    const btype = m[1].toLowerCase();
    let start = i + m[0].length - 1;
    let depth = 0;
    let j = start;
    for (; j < s.length; j++) {
      if (s[j] === "{") depth++;
      else if (s[j] === "}") {
        depth--;
        if (depth === 0) {
          blocks.push([btype, s.slice(start + 1, j).trim()]);
          i = j + 1;
          break;
        }
      }
    }
    if (j >= s.length) break;
  }
  return blocks;
}

function convertScript(script) {
  const lines = [];
  const counters = { stage: 0, filter: 0, comp: 0, env: 0, osc: 0 };
  for (const [btype, body] of extractBlocks(script)) {
    if (btype === "param") {
      const kv = parseKvBody(body);
      lines.push(`param ${kv.alias ?? "a"} = ${kv.name ?? kv.alias ?? "a"} [${kv.min ?? "0"}, ${kv.max ?? "1"}]`);
    } else if (btype === "stage") {
      counters.stage++;
      const compact = body.replace(/\s+/g, " ").trim();
      const formula = compact.startsWith("y ") ? compact.slice(2).trim() : compact;
      lines.push(`stage${counters.stage}: y = ${formula}`);
    } else if (btype === "filter") {
      counters.filter++;
      const kv = parseKvBody(body);
      lines.push(`filter${counters.filter}: ` + Object.entries(kv).map(([k, v]) => `${k} = ${v}`).join("; "));
    } else if (btype === "osc") {
      const kv = parseKvBody(body);
      const name = kv.name ?? `osc${counters.osc + 1}`;
      counters.osc++;
      const shape = shapeMap[kv.shape?.toLowerCase()] ?? kv.shape ?? "sine";
      const parts = [`shape = ${shape}`];
      if (kv.freq) parts.push(`freq = ${kv.freq}`);
      if (kv.sync) parts.push(`sync = ${kv.sync}`);
      if (kv.depth) parts.push(`depth = ${kv.depth}`);
      lines.push(`${name}: ` + parts.join("; "));
    } else if (btype === "env") {
      const kv = parseKvBody(body);
      const name = kv.name ?? `env${counters.env + 1}`;
      counters.env++;
      const parts = [`type = ${(kv.mode ?? "rms").toLowerCase()}`];
      if (kv.attack) parts.push(`attack = ${kv.attack}`);
      if (kv.release) parts.push(`release = ${kv.release}`);
      if (kv.trigger) parts.push(`trigger = ${kv.trigger}`);
      lines.push(`${name}: ` + parts.join("; "));
    } else if (btype === "comp") {
      counters.comp++;
      const kv = parseKvBody(body);
      const parts = ["threshold", "ratio", "attack", "release"].filter((k) => kv[k]).map((k) => `${k} = ${kv[k]}`);
      lines.push(`comp${counters.comp}: ` + parts.join("; "));
    }
  }
  return lines.join("\n");
}

const file = path.join(__dirname, "..", "resources", "factory_presets.json");
const data = JSON.parse(fs.readFileSync(file, "utf8"));
for (const preset of data) preset.script = convertScript(preset.script);
fs.writeFileSync(file, JSON.stringify(data, null, 2) + "\n", "utf8");
console.log(`Converted ${data.length} presets`);